### ISSUE TITLE
whois auth information returns incorrect results

### ISSUE THREAD

**Issue by eklitzke** [oragono/oragono#217]
The way I expect this to work is as it does on Freenode. On Freenode the account I'm registered with is `eck`, but I also use the handle `eklitzke` in some development channels, and I've linked my `eklitzke` account to the `eck` account.

I just logged in using the Freenode webchat client, and here's what I see:

```
/whois eck
[17:22] == eck [~postfix@fsf/member/eck]
[17:22] ==  realname : sendmail
[17:22] ==  server   : moon.freenode.net [Atlanta, GA, US]
[17:22] ==           : is using a secure connection
[17:22] ==  account  : eck
[17:22] == End of WHOIS

/whois eklitzke
[17:22] == eklitzke [~evan@fsf/member/eck]
[17:22] ==  realname : Evan Klitzke
[17:22] ==  server   : hitchcock.freenode.net [Sofia, BG, EU]
[17:22] ==           : is using a secure connection
[17:22] ==  account  : eck
[17:22] == End of WHOIS
```

As you can see, they both show me authed as `eck`. This would also apply if I used `/nick` on one of those accounts, it would still link me back to my `eck` handle.

Here's what I actually see right now on `irc.darwin.network` (where I am logged in as `ghost` while I was testing some `/umode` stuff).

```
/whois ghost
17:24 -- [ghost] (~death@hacker.monad.io): death
17:24 -- [ghost] #darwin
17:24 -- [ghost] 75.101.96.6 :Actual user@host, Actual IP ~death@hacker.monad.io
17:24 -- [ghost] is using a secure connection
17:24 -- [evan] logged in as :is
17:24 -- ghost en :can speak these languages
17:24 -- [ghost] idle: 00 hours 06 minutes 09 seconds, signon at: Thu, 15 Mar 2018 17:10:54
17:24 -- [ghost] End of /WHOIS list
```

OK so far so good, I am actually authed as `evan` and that shows up correctly (although the formatting is terrible, see below). Let me try looking up another user:
```
/whois nish
17:25 -- [nis] End of /WHOIS list
17:25 -- [nish] (~nishbot@198.199.114.107): Supybot 0.83.4.1
17:25 -- [nish] #darwin
17:25 -- [nish] is using a secure connection
17:25 -- nish en :can speak these languages
17:25 -- [nish] idle: 00 hours 02 minutes 12 seconds, signon at: Thu, 15 Mar 2018 01:25:38
17:25 -- [nish] End of /WHOIS list
```

That also looks OK. But I see strange results when I `/whois` this guy who previously tried to steal my nick:
```
17:26 -- [shivaram] (~shivaram@c-107-3-81-17.hsd1.ma.comcast.net): shivaram
17:26 -- [shivaram] #darwin
17:26 -- [shivaram] is using a secure connection
17:26 -- [evan] logged in as :is
17:26 -- shivaram en :can speak these languages
17:26 -- [shivaram] idle: 00 hours 02 minutes 46 seconds, signon at: Wed, 14 Mar 2018 06:57:00
17:26 -- [shivaram] End of /WHOIS list
```

This shivaram guy shows up being authed as me, which is not what I expect. I'd like to ghost him, but when I `/msg nickserv shivaram` it doesn't work:

```
17:27 -- MSG(nickserv): ghost shivaram
17:27 -- irc.darwin.network: You don't own that nick
```

Indeed I do not own that nick, so it seems like the ircd has the right mapping but is returning bad whois data.

It also looks like the formatting isn't right, the lines look like

```
17:26 -- [evan] logged in as :is
```
but I believe it should say
```
17:26 -- :is logged in as [evan]
```

My guess is that some prankster was using RTL characters while impersonating me. It's still strange that would affect /whois on my own account.

**Comment by slingamn** [comment]
The core bug is pretty straightforward, here's a fix:

https://github.com/slingamn/oragono/commit/74c243d5aeab05820eea8d5844e2fec66bb56262

however, clients still may not be interpreting the resulting output correctly. In Hexchat raw I/O:

````
<< whois netcat
>> @time=2018-03-16T00:42:46.892Z :oragono.test 311 shivaram_hexchat netcat ~netcat 0::1 * netcat
>> @time=2018-03-16T00:42:46.892Z :oragono.test 319 shivaram_hexchat netcat :@#shivaramtest 
>> @time=2018-03-16T00:42:46.892Z :oragono.test 330 shivaram_hexchat netcat :is logged in as
>> @time=2018-03-16T00:42:46.892Z :oragono.test 317 shivaram_hexchat netcat 25 1521160924 :seconds idle, signon time
>> @time=2018-03-16T00:42:46.892Z :oragono.test 318 shivaram_hexchat netcat :End of /WHOIS list
````

but in the UI:

````
* [netcat] (~netcat@0::1): netcat
* [netcat] @#shivaramtest 
* [netcat] ogged in as :is
* [netcat] idle 00:00:25, signon: Thu Mar 15 20:42:04
* [netcat] End of WHOIS list.
````

If I Google this, I find a number of related client bugs:

* https://bugs.quassel-irc.org/issues/1145
* https://forums.mirc.com/ubbthreads.php/topics/243026/debug_window_330_numeric_issue

**Comment by slingamn** [comment]
Another one: https://bugzilla.mozilla.org/show_bug.cgi?id=217474

**Comment by eklitzke** [comment]
According to https://www.alien.net.au/irc/irc2numerics.html it is not recommended to use a 330 response code as it has conflicting definitions, but I'm not sure how official of a resource this is.

**Comment by DanielOaks** [comment]
Well, 'user' is a specific part of the nickmask, `nick!user@host`, so changing 311 to return the account name instead of the user part of the nickmask would mess with things a fair bit more. It'd be worth checking with the IRCds around and seeing whether any (and how many) servers still use 330 as `RPL_WHOWAS_TIME`. After we get those details, we can file updates against the [irc-defs](https://github.com/ircdocs/irc-defs) [numerics list](https://defs.ircdocs.horse/defs/numerics.html) (the updated fork of the alien.net.au lists).

**Comment by eklitzke** [comment]
I saw some discussion about this topic in #ircv3 today (well related, it was about general rpc-keyvalue responses in whois queries). Seems like this is on track to be fixed in general in IRC One Day.

**Comment by slingamn** [comment]
Weechat issue link for reference: weechat/weechat#1160

**Comment by slingamn** [comment]
Oragono did in fact have a bug in its `RPL_WHOISACCOUNT` response line; fixed in #290.
